osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/Sync/CatchUpSyncManager.cs

```diff
@@ -0,0 +1,150 @@
+        /// </summary>
+        private readonly List<ISpectatorPlayerClock> playerClocks = new List<ISpectatorPlayerClock>();
+
+        private bool hasStarted;
+        private double? firstStartAttemptTime;
+
+        public CatchUpSyncManager(IAdjustableClock master)
+        {
+            MasterClock = master;
+        }
+
+        public void AddPlayerClock(ISpectatorPlayerClock clock) => playerClocks.Add(clock);
+
+        public void RemovePlayerClock(ISpectatorPlayerClock clock) => playerClocks.Remove(clock);
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (!attemptStart())
+            {
+                // Ensure all player clocks are stopped until the start succeeds.
+                foreach (var clock in playerClocks)
+                    clock.Stop();
+                return;
+            }
+
+            updateCatchup();
+            updateMasterClock();
+        }
+
+        /// <summary>
+        /// Attempts to start playback. Waits for all player clocks to have available frames for up to <see cref="MAXIMUM_START_DELAY"/> milliseconds.
+        /// </summary>
+        /// <returns>Whether playback was started and syncing should occur.</returns>
+        private bool attemptStart()
+        {
+            if (hasStarted)
+                return true;
+
+            if (playerClocks.Count == 0)
+                return false;
+
+            int readyCount = playerClocks.Count(s => !s.WaitingOnFrames.Value);
+
+            if (readyCount == playerClocks.Count)
+                return hasStarted = true;
+
+            if (readyCount > 0)
+            {
+                firstStartAttemptTime ??= Time.Current;
+
+                if (Time.Current - firstStartAttemptTime > MAXIMUM_START_DELAY)
+                    return hasStarted = true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Updates the catchup states of all player clocks clocks.
+        /// </summary>
+        private void updateCatchup()
+        {
+            for (int i = 0; i < playerClocks.Count; i++)
+            {
+                var clock = playerClocks[i];
+                double timeDelta = MasterClock.CurrentTime - clock.CurrentTime;
+
+                // Check that the player clock isn't too far ahead.
+                // This is a quiet case in which the catchup is done by the master clock, so IsCatchingUp is not set on the player clock.
+                if (timeDelta < -SYNC_TARGET)
+                {
+                    clock.Stop();
+                    continue;
+                }
+
+                // Make sure the player clock is running if it can.
+                if (!clock.WaitingOnFrames.Value)
+                    clock.Start();
+
+                if (clock.IsCatchingUp)
+                {
+                    // Stop the player clock from catching up if it's within the sync target.
+                    if (timeDelta <= SYNC_TARGET)
```

Comment:
I think we're going to want to set this to `SYNC_TARGET / 2`, else there may be a case of the state flip-flopping between catch-up and non-catchup.